src/home/homework.c

```diff
@@ -1795,8 +1799,43 @@ static int home_inspect(UserRecord *h, UserRecord **ret_home) {
         return 1;
 }
 
+static int user_session_freezer(UserRecord *h, bool freeze, UnitFreezer *ret) {
```

Comment:
Hmm another issue: this freeze in homed in combination w/ systemd-sleep will thaw the homed freeze early

Test I just ran on my machine:
```
# systemctl freeze user@60071.service
# systemctl freeze user.slice
# systemctl status user@60071.service
(indicates that it's frozen)
# systemctl status user.slice
(indicates that it's frozen)
# systemctl thaw user.slice
# systemctl status user.slice
(indicates that it's no longer frozen. Expected)
# systemcl status user@60071.service
(indicates that it's no longer frozen. UNEXPECTED!)
```

I feel that this behavior is wrong. It also negates the benefits of freezing the homed user session:

- User closes lid
- gnome-shell decides that it wants to drop encryption keys. So, it switches to GDM and calls Lock on the home dir
- homed freezes the user session, to protect against the aforementioned edge cases, then drops the encryption keys and freezes the LUKS device
- systemd-sleep now executes. It freezes all user.slice to freeze all user sessions
- system goes into suspend
- (some time later) system is resumed
- systemd-sleep continues execution. It thaws user.slice
- This now thaws the homed user session too!
- gnome-shell and the rest of the user session are running again. But file I/O is still blocked! Queue all of the edge cases I mentioned in a previous comment
- (some time later) the user enters their password. This unlocks the LUKS device and the danger has passed.

There can be an arbitrary amount of time between the resume and the user entering their password. Thus, we can have any of the edge cases happen and lead to data loss.

@poettering any insight on this? Why is it behaving this way? Anything we can do about it?

Reply:
Quick skim through the code: I don't think this is a limitation of cgroups; it's just how freeze/thaw is implemented on slices. Right now freezing a slice recursively freezes all children, and thawing a slice recursively thaws all children. No consideration is paid to what state the children were in before the freeze or thaw.

Perhaps we should have a refcount of how many levels of freeze a unit is in, so that every freeze should be paired with a thaw for the service to continue executing?

Reply:
Looked @ kernel documentation for cgroup freezing (for cgroup v2) and am not a bit more confused why systemd implements this the way it does. According to the docs, freezing a cgroup implicitly freezes all decedents. In other words, if I freeze a slice, everything below that slice should be frozen too.

So my updated proposal on what systemd should be doing: when you freeze a slice, it should freeze the one slice and that's it. When you thaw the slice, the same thing should happen. The kernel handles freezing/thawing all descendant slices for systemd. Thus (if I'm understanding the kernel docs correctly!) I could freeze a child slice, then a parent, then thaw the parent and have the child remain frozen.

Reply:
I moved this out into https://github.com/systemd/systemd/issues/30640 so resolving here